### PR TITLE
ENH: Moves calling of TradingAlgorithm.initialize to _create_generator

### PR DESCRIPTION
**Pull request by jfkirk**
This commit moves the call to the algo code's initialize method out of TradingAlgorithm.__init__. The algo code's initialize method will be called in _create_generator using args and kwargs that are stored from __init__.

### CODE REVIEW

**Comment by jfkirk** [comment]
@ssanderson Could you review this? If you're overloaded I can find a different reviewer. It was a straightforward change to move calling the algocode's initialize method.

File: zipline/algorithm.py [hunk 1]
```diff
@@ -248,13 +244,10 @@ def __init__(self, *args, **kwargs):
...
 
-        # Subclasses that override initialize should only worry about
-        # setting self.initialized = True if AUTO_INITIALIZE is
-        # is manually set to False.
+        # Prepare the algo for initialization
         self.initialized = False
-        self.initialize(*args, **kwargs)
-        if self.AUTO_INITIALIZE:
-            self.initialized = True
+        self.initialize_args = args
```
Comment (ssanderson) [inline zipline/algorithm.py]: What code do we have that depends on `initialize` receiving the extra `*args` and `**kwargs`?  Can we remove this give `TradingAlgorithm.__init__` a fixed signature?  :IWouldBeSoHappy: if we could do that.

File: zipline/algorithm.py [hunk 2]
```diff
@@ -371,6 +364,11 @@ def _create_generator(self, sim_params, source_filter=None):
         processed by the zipline, and False for those that should be
         skipped.
         """
+
+        if not self.initialized:
+            self.initialize(*self.initialize_args, **self.initialize_kwargs)
```
Comment (ssanderson) [inline zipline/algorithm.py]: See note above re: these parameters.

File: zipline/algorithm.py (hunk 1 above)
Comment (jfkirk) [inline zipline/algorithm.py]: The initialize methods in AlgorithmProxy and LiveSessionAlgorithmProxy have different signatures. I will chalk it on to the forthcoming Zipline initialization cleanup/refactoring.

**Comment by jfkirk** [comment]
@StewartDouglas @yankees714 For more eyes

**Comment by twiecki** [comment]
Awesome! This enables parameter optimization :+1:

**Comment by warren-oneill** [comment]
@twiecki oh didn't realize that! great!

**Comment by femtotrader** [comment]
About parameter optimization

this https://jug.readthedocs.org/en/latest/ might also help

I have seen it in "building machine learning systems with python - second edition"

**Pull request merged**